### PR TITLE
Auth0

### PR DESCRIPTION
**Pull request by carvilsi**
- adding auth basic to sdk logIn
- users tests
- adding lokiJs

### CODE REVIEW

File: src/CoreManager.coffee [hunk 1]
```diff
@@ -17,5 +20,11 @@ class CoreManager
 
   executeOperations: (operations) ->
     @commController.write(operations)
+    
+  getDB: ->
```
Comment (malamili) [inline src/CoreManager.coffee]: More consistent to name this getUsersDB like you did in Weaver.coffee

File: src/WeaverUser.coffee [hunk 2]
```diff
@@ -1,26 +1,54 @@
...
+      password:pass
+    }
+    coreManager = Weaver.getCoreManager()
+    users = Weaver.getUsersDB()
+    coreManager.logIn(credentials)
+    .then((res) ->
+      try
+        users.insert({user:usr,token:res[0].token})
+      catch error
+        console.error(error)
```
Comment (malamili) [inline src/WeaverUser.coffee]: What error do you expect here? The user of the SDK might not get feedback of this error when printed. Is it not better to either let it fail or reject with some error code

File: test/WeaverUser.test.coffee [hunk 3]
```diff
@@ -17,50 +20,85 @@ describe 'Weaver User', ->
     return
 
 
-  it 'should signup users', (done) ->
-    Weaver.User.signUp('asdf', 'zxcv')
-    .then((user) =>
-      assert(user.getSessionToken())
-      done()
-    )
+  # it 'should signup users', (done) ->
```
Comment (malamili) [inline test/WeaverUser.test.coffee]: This is still work in progress?

File: test/WeaverUser.test.coffee [hunk 4]
```diff
@@ -4,10 +4,13 @@ require("./test-suite")()
 Weaver      = require('./../src/Weaver')
 WeaverError = require('./../../weaver-commons-js/src/WeaverError')
 require('./../src/WeaverNode')  # This preloading will be an issue
+require('./../src/WeaverUser')
 
 describe 'Weaver User', ->
+  this.timeout(2000)
 
   before (done) ->
+    console.log WEAVER_ADDRESS
```
Comment (malamili) [inline test/WeaverUser.test.coffee]: Please remove console.log statements

File: test/WeaverUser.test.coffee [hunk 5]
```diff
@@ -17,50 +20,85 @@ describe 'Weaver User', ->
...
-      Weaver.User.signUp('asdf', 'asdf3')
-    ).then().catch((error) ->
-      assert.equal(error.code, WeaverError.USERNAME_TAKEN)
-      done()
+    
+  it 'should fails when trying to login with non existing user', ->
+    user = new Weaver.User()
+    user.logIn('andromeda','Chains')
+    .then().catch((err) ->
+      console.log err
```
Comment (malamili) [inline test/WeaverUser.test.coffee]: In stead of console.log, use the chai as promissed to expect a reject (or use done() in the catch block)

File: src/CoreManager.coffee (hunk 1 above)
Comment (carvilsi) [inline src/CoreManager.coffee]: 👍

File: test/WeaverUser.test.coffee (hunk 4 above)
Comment (carvilsi) [inline test/WeaverUser.test.coffee]: 👍

File: test/WeaverUser.test.coffee (hunk 3 above)
Comment (carvilsi) [inline test/WeaverUser.test.coffee]: Yep, I just leave this block commented, by now. I will uncomment when I'll finish this part

File: test/WeaverUser.test.coffee (hunk 5 above)
Comment (carvilsi) [inline test/WeaverUser.test.coffee]: Thanks

File: test/WeaverUser.test.coffee [hunk 6]
```diff
@@ -16,51 +18,100 @@ describe 'Weaver User', ->
...
+          .then((res, err) ->
+            res.token.should.be.a('string')
+          )
+        )
+      )
+    )
+    
+  it 'should signOff a user', ->
+    user = new Weaver.User()
+    user.current('centaurus').then((res) ->
```
Comment (malamili) [inline test/WeaverUser.test.coffee]: This is bad chaining of promises, use it as follows:

```
then(->
usr.logOut()
).then(->
user.login('')
)
.then(->
res.should.be...
)
```

File: src/CoreManager.coffee [hunk 7]
```diff
@@ -17,5 +20,17 @@ class CoreManager
...
+  getUsersDB: ->
+    @users
+    
+  logIn: (credentials) ->
+    @commController.logIn(credentials)
+    
+  signUp: (newUserPayload) ->
+    @commController.signUp(newUserPayload)
+    
+  signOff: (userPayload) ->
```
Comment (malamili) [inline src/CoreManager.coffee]: Better to be consistent and call this logOut too

Comment (malamili) [inline src/CoreManager.coffee]: ...even throughout the server emit messages and controllers

File: src/WeaverProject.coffee [hunk 8]
```diff
@@ -4,7 +4,7 @@ WeaverNode = require('./WeaverNode')
 
 
 module.exports =
-  class WeaverUser extends WeaverNode
+  class WeaverProject extends WeaverNode
```
Comment (malamili) [inline src/WeaverProject.coffee]: revert this change, not relevant and will cause unneccesairy conflicts with gijs his coming pr

File: src/WeaverUser.coffee [hunk 9]
```diff
@@ -1,26 +1,98 @@
 # Libs
-cuid = require('cuid')
-WeaverNode = require('./WeaverNode')
+cuid   = require('cuid')
+Weaver = require('./Weaver')
+loki   = require('lokijs')
 
+class WeaverUser
+
```
Comment (malamili) [inline src/WeaverUser.coffee]: put these three fields in the constructor

Comment (malamili) [inline src/WeaverUser.coffee]: (email, username, password etc)

File: src/WeaverUser.coffee [hunk 10]
```diff
@@ -1,26 +1,98 @@
...
+  
+  constructor: () ->
+    @email
+    @username
+    @password
+    @emailVerified = false
+    
+    
+  logIn: (usr, pass) ->
+    credentials = {
```
Comment (malamili) [inline src/WeaverUser.coffee]: I'll comment it here but it applies for lots of code below..if the key name is the same as the value name, please write like this:

`credentials = {user,pass}`

this would require to rename usr to user, but that is better anyways

File: src/WeaverUser.coffee [hunk 11]
```diff
@@ -1,26 +1,98 @@
...
+      try
+        #
+        users.insert({user:usr,token:res.token})
+      catch error
+        console.error(error)
+      res
+    )
+  
+  signUp: (currentUsr,userName,userEmail,userPassword,directoryName) ->
+    newUserCredentials = {
```
Comment (malamili) [inline src/WeaverUser.coffee]: so here too like above

File: src/WeaverUser.coffee [hunk 12]
```diff
@@ -1,26 +1,98 @@
...
+    credentials = {
+      user: usr
+      password: pass
+    }
+    coreManager = Weaver.getCoreManager()
+    users = Weaver.getUsersDB()
+    coreManager.logIn(credentials)
+    .then((res) ->
+      try
+        #
```
Comment (malamili) [inline src/WeaverUser.coffee]: hash?

File: src/WeaverUser.coffee [hunk 13]
```diff
@@ -1,26 +1,98 @@
...
+        users.insert({user:usr,token:res.token})
+      catch error
+        console.error(error)
```
Comment (malamili) [inline src/WeaverUser.coffee]: this is bad practice, throw an error or let it fail, but do not log this

File: src/WeaverUser.coffee [hunk 14]
```diff
@@ -1,26 +1,98 @@
...
+    
+    
+  # Returns the "valid" token from current user, null if not loggedin
+  current: (usr) ->
+    new Promise((resolve, reject) =>
+      try
+        users = Weaver.getUsersDB()
+        resolve(users.findOne({user:usr}).token)
+      catch error
+        reject(null)
```
Comment (malamili) [inline src/WeaverUser.coffee]: why not reject with error message?

File: src/WeaverUser.coffee [hunk 15]
```diff
@@ -1,26 +1,98 @@
...
+    new Promise((resolve, reject) =>
+      try
+        users = Weaver.getUsersDB()
+        if usr?
+          users.remove(users.find({$and:[{user:usr},{token:{$ne:undefined}}]}))
+          resolve()
+        else
+          users.remove(users.find({token:{$ne:undefined}}))
+          resolve()
+      catch error
```
Comment (malamili) [inline src/WeaverUser.coffee]: could you give an example of this catch?

File: test/WeaverUser.test.coffee [hunk 16]
```diff
@@ -16,51 +18,100 @@ describe 'Weaver User', ->
...
-    Weaver.User.signUp('asdf', 'zxcv')
-    .then((user) =>
-      assert(user.getSessionToken())
+  it 'should login users, receiving a valid jwt', ->
+    user = new Weaver.User()
+    user.logIn('phoenix','phoenix')
+    .then((res) ->
+      token = res.token
+      token.should.be.a('string')
+    ).catch((err) ->
```
Comment (malamili) [inline test/WeaverUser.test.coffee]: this catch in the test is not used, please remove

File: test/WeaverUser.test.coffee [hunk 17]
```diff
@@ -16,51 +18,100 @@ describe 'Weaver User', ->
...
+    )
+    
+  it 'should fails when trying to login with non existing user', ->
+    user = new Weaver.User()
+    user.logIn('andromeda','Chains')
+    .then().catch((err) ->
       done()
```
Comment (malamili) [inline test/WeaverUser.test.coffee]: can you assert on the error? see WeaverNode.test for an exampleo

File: test/WeaverUser.test.coffee [hunk 18]
```diff
@@ -16,51 +18,100 @@ describe 'Weaver User', ->
...
-
-  it 'should login users', (done) ->
-    Weaver.User.signUp('asdf', 'zxcv')
-    .then(->
-      Weaver.User.logIn('asdf', 'zxcv')
+  
+  
+  it 'should returns jwt from the loggedin user', ->
+    user = new Weaver.User()
+    user.current('phoenix')
```
Comment (malamili) [inline test/WeaverUser.test.coffee]: if you use chai as promised, you can write:

`user.current('phoenix').should.eventually.be.a.('string')`

File: test/WeaverUser.test.coffee (hunk 16 above)
Comment (carvilsi) [inline test/WeaverUser.test.coffee]: 👍

File: src/WeaverProject.coffee (hunk 8 above)
Comment (carvilsi) [inline src/WeaverProject.coffee]: 👍

File: src/CoreManager.coffee (hunk 7 above)
Comment (carvilsi) [inline src/CoreManager.coffee]: I do not understand what's the point here. Please can you explain better? SignOff is "deleting" the user from the platform and logOut is just leave the platform for the moment.

File: src/WeaverUser.coffee (hunk 9 above)
Comment (carvilsi) [inline src/WeaverUser.coffee]: 👍

File: src/CoreManager.coffee (hunk 7 above)
Comment (malamili) [inline src/CoreManager.coffee]: oh sorry, didnt know we support this already! great then never mind 👍

File: src/WeaverUser.coffee (hunk 12 above)
Comment (carvilsi) [inline src/WeaverUser.coffee]: Is not a hash is a feature.

File: src/WeaverUser.coffee (hunk 13 above)
Comment (carvilsi) [inline src/WeaverUser.coffee]: Indeed! Thanks.

File: src/WeaverUser.coffee (hunk 10 above)
Comment (carvilsi) [inline src/WeaverUser.coffee]: Yeah, so much better.

File: src/WeaverUser.coffee (hunk 11 above)
Comment (carvilsi) [inline src/WeaverUser.coffee]: 👍

File: test/WeaverUser.test.coffee (hunk 6 above)
Comment (carvilsi) [inline test/WeaverUser.test.coffee]: 👍

File: src/CoreManager.coffee (hunk 7 above)
Comment (carvilsi) [inline src/CoreManager.coffee]: 👍

File: src/WeaverUser.coffee (hunk 15 above)
Comment (carvilsi) [inline src/WeaverUser.coffee]: nop… but I'm interacting with the in memory database and I'm worry about any fail hence the try-catch.
What I should do?

File: test/WeaverUser.test.coffee (hunk 18 above)
Comment (carvilsi) [inline test/WeaverUser.test.coffee]: Cool!

File: src/WeaverUser.coffee (hunk 14 above)
Comment (carvilsi) [inline src/WeaverUser.coffee]: This is function will not be exposed (at least is this my idea) to the user of the SDK so I decided to reject with null.

File: test/WeaverUser.test.coffee (hunk 17 above)
Comment (carvilsi) [inline test/WeaverUser.test.coffee]: Yes, I will. When I have the filtering errors at the weaver server.
Getting this at the moment: 
```
{ name: 'StatusCodeError',
  statusCode: 404,
  message: '404 - {"Error":{"code":404,"message":"Document not Found"}}',
  error: { Error: { code: 404, message: 'Document not Found' } },
  options: 
   { method: 'GET',
     uri: 'http://*******/api/v1/token',
     headers: { Authorization: 'Basic YW5kcm9tZWRhOkNoYWlucw==' },
     json: true,
     simple: true,
     resolveWithFullResponse: false,
     transform2xxOnly: false },
  response: 
   { statusCode: 404,
     body: { Error: [Object] },
     headers: 
      { date: 'Thu, 12 Jan 2017 10:45:29 GMT',
        'access-control-allow-origin': '*',
        'access-control-request-method': '*',
        'access-control-allow-headers': '*',
        'content-type': 'application/json',
        connection: 'close',
        server: 'Jetty(9.3.z-SNAPSHOT)' },
     request: { uri: [Object], method: 'GET', headers: [Object] } } }
```
